### DISCUSSION TITLE
User friendly error messages?

### DISCUSSION THREAD

**Discussion by laneme** [colinhacks/zod#1328] [Q&A]
Consider the schema below 
```ts
  const res = object({
      displayName: string().min(2).max(30),
  }).safeParse(values) // values = {displayName: "A"}
```
This returns an issue containing `String must contain at least 2 character(s)`. But when i'm using them with forms, i need to display the error messages to users. They might not know what is a string.

An ideal error message would've been `Display Name must contain at least 2 characters(s)`, where i can pass the label (Display Name) like
```ts
displayName: string().min(2).label('Display Name')
```
Is there already a way to do this with zod? Passing an error message manually to every schema property just increases my code and takes away the visual simplicity zod provides me. Thank you

**Comment by nullndr** [comment]
Sure, you just have to use the [`ZodErrorMap`](https://github.com/colinhacks/zod/blob/master/ERROR_HANDLING.md#customizing-errors-with-zoderrormap).

A simple examples for your case if the following:

```ts
const errorMap: z.ZodErrorMap = (error, ctx) => {
  switch (error.code) {
    case z.ZodIssueCode.too_small:
      return { message: "Display Name must contain at least 2 characters" };
  }

  return { message: ctx.defaultError };
};

const myString = z
  .string({
    errorMap,
  })
  .min(2)
  .max(30);

const data = myString.safeParse("a");

if (data.success) {
  console.log(data.data);
} else {
  console.log(data.error.issues);
}
```

This produces the following:

```
[
  {
    code: 'too_small',
    minimum: 2,
    type: 'string',
    inclusive: true,
    message: 'Display Name must contain at least 2 characters',
    path: []
  }
]
```
